packages/studio-ui/src/web/components/Layout/TopNav/cloud/Step3-deploy.tsx

```diff
@@ -0,0 +1,152 @@
+import axios from 'axios'
+import { isCDMError, UnreachableCaseError } from 'common/errors'
+import React, { useEffect, useReducer } from 'react'
+import { connect } from 'react-redux'
+import { Timeout, toastFailure } from '~/components/Shared/Utils'
+import { RootReducer } from '~/reducers'
+import { Status } from './deploy'
+import { TrainSessions } from './types'
+
+const BASE_NLU_URL = `${window.STUDIO_API_PATH}/nlu`
+
+const computeIsTrained = (trainSessions: TrainSessions) =>
+  Object.values(trainSessions).reduce((trained, session) => trained && session.status === 'done', true)
+
+type State =
+  | {
+      status: 'train_pending'
+    }
+  | { status: 'training_in_progress' }
+  | { status: 'upload_pending' }
+  | { status: 'upload_completed' }
+  | { status: 'upload_failed' }
+
+type Action =
+  | { type: 'training/started' }
+  | { type: 'training/trainSessionsUpdated'; trainSessions: TrainSessions }
+  | { type: 'upload/ended' }
+  | { type: 'upload/failed' }
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'training/started':
+      return { status: 'training_in_progress' }
+    case 'training/trainSessionsUpdated':
+      const isTrained = computeIsTrained(action.trainSessions)
+      if (isTrained) {
+        return { status: 'upload_pending' }
+      } else {
+        return { status: 'training_in_progress' }
+      }
+    case 'upload/ended':
+      return { status: 'upload_completed' }
+    case 'upload/failed':
+      return { status: 'upload_completed' }
+    default:
+      throw new UnreachableCaseError(action)
+  }
+}
+
+interface OwnProps {
+  pat: string
+  workspaceId: string
+  onCompleted: () => void
+}
+
+type StateProps = ReturnType<typeof mapStateToProps>
+type DispatchProps = typeof mapDispatchToProps
+type Props = DispatchProps & StateProps & OwnProps
+
+function init(props: Props): State {
+  const { trainSessions } = props
+
+  const isTrained = computeIsTrained(trainSessions)
+
+  if (isTrained) {
+    return { status: 'upload_pending' }
+  } else {
+    return { status: 'train_pending' }
+  }
+}
+
+export const Deploy = (props: Props): JSX.Element => {
+  const { pat, trainSessions, workspaceId, onCompleted } = props
+
+  const [state, dispatch] = useReducer(reducer, props, init)
+
+  const { status } = state
+
+  useEffect(() => {
+    const ac = new AbortController()
```

Comment:
**nitpick (non-blocking):** seems like you're using this pattern quite often, you could create a reusable react hook for it instead

Reply:
I'll try this
```
function abortController() {
  const ac = new AbortController()
  useEffect(() => {
    return () => {
      // Cancel all pending requests
      ac.abort()
    }
  }, [])
  
  return axios.createInstance({signal: ac})
}

const axios = useAbortedAxios() // axios has
```

Reply:
this can be done another time, really not salty about this